File: deno/day02.ts
import { type DaySolution, parseExempleData, parseInputPart1Data } from './_tools.ts';
import { parseNumbersByLines } from './_helpers.ts';

function isSafe(report: number[]): boolean {
  const increase = report[1] - report[0] > 0;
  return report
    .slice(0, -1)
    .every((_, i) => {
      const diff = report[i + 1] - report[i];
      return Math.abs(diff) >= 1 &&
        Math.abs(diff) <= 3 &&
        increase === diff >= 0;
    });
}

export class Day02 implements DaySolution<number[][], number> {
  day = 2;

  parseData(dataLabel: 'Example' | 'Input'): number[][] {
    const rawData = dataLabel === 'Example' ? parseExempleData(this.day) : parseInputPart1Data(this.day);

    return parseNumbersByLines(rawData);
  }

  part1(reports: number[][]): number {
    return reports.filter(isSafe).length;
  }

  part2(reports: number[][]): number {
    return reports
      .map((report) => report.map((_, i) => report.filter((_, j) => j !== i)))
      .filter((subReports) => subReports.some(isSafe))
      .length;
  }
}

File: rust/src/day02.rs
use crate::tools::{read_example, read_input};
use crate::DaySolution;

pub struct Day02 {}

impl Day02 {
    fn is_safe(&self, report: &Vec<i32>) -> bool {
        let increase = report[1] - report[0] >= 0;
        report.windows(2)
            .all(|pair| {
                let diff = pair[1] - pair[0];
                let pair_increase = diff >= 0;
                increase == pair_increase && diff.abs() <= 3 && diff.abs() >= 1
            })
    }
}

impl DaySolution<Vec<Vec<i32>>, i32> for Day02 {
    fn day(&self) -> u16 {
        2_u16
    }

    fn parse_input(&self, example: bool) -> Vec<Vec<i32>> {
        let raw_data = if example {
            read_example(self.day())
        } else {
            read_input(self.day())
        };

        raw_data.lines()
            .filter(|line| !line.is_empty())
            .map(|line| {
                line.split_whitespace()
                    .map(|x| x.parse().unwrap())
                    .collect()
            })
            .collect()
    }

    fn part1(&self, input: &Vec<Vec<i32>>) -> i32 {
        input.iter()
            .map(|report| self.is_safe(&report))
            .filter(|&safe| safe)
            .count() as i32
    }

    fn part2(&self, input: &Vec<Vec<i32>>) -> i32 {
        input.iter()
            .map(|report| {
                (0..report.len())
                    .map(|i| {
                        let mut new_report = report.clone();
                        new_report.remove(i);
                        new_report
                    })
                    .any(|report| self.is_safe(&report))
            })
            .filter(|&safe| safe)
            .count() as i32
    }
}
